### PR TITLE
[ISSUE #3929] wake up client’s waiting when server crashed

### PR DESCRIPTION
**Pull request by qsrg**
This PR is aimed to  wake up client’s waiting when server crashed. 
When producer has send message to broker successfully and wait for response，but broker crashed at this time，the produer would  blocked until timeout and wouldn't retry. This PR can detect broker crashed and wake up producer's waiting state, then producer will resend message to other broker.

### CODE REVIEW

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java [hunk 1]
```diff
@@ -707,5 +711,25 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
                 NettyRemotingClient.this.putNettyEvent(new NettyEvent(NettyEventType.EXCEPTION, remoteAddress, ctx.channel()));
             }
         }
+
+        @Override
+        public void channelInactive(ChannelHandlerContext ctx) throws Exception {
```
Comment (odbozhou) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java:716]: Can channelInactive represent broker crash?

Comment (qsrg) [reply remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java:716]: Usually, it does work

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java [hunk 2]
```diff
@@ -38,12 +38,7 @@
 import java.io.IOException;
 import java.net.SocketAddress;
 import java.security.cert.CertificateException;
-import java.util.Collections;
-import java.util.List;
-import java.util.Map;
-import java.util.Random;
-import java.util.Timer;
-import java.util.TimerTask;
+import java.util.*;
```
Comment (RongtongJin) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java:41]: Do not import xxx.*

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java [hunk 3]
```diff
@@ -707,5 +711,25 @@ public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws E
...
+                    Map.Entry<Integer, ResponseFuture> entry = it.next();
+                    if (channel == entry.getValue().getProcessChannel()) {
+                        final String remoteAddress = RemotingHelper.parseChannelRemoteAddr(ctx.channel());
+                        log.warn("NETTY CLIENT PIPELINE: channelInactive, so wakeup response, the channel[{}]", remoteAddress);
+                        entry.getValue().wakeupResponse();
+                    }
+                }
+            }
+            super.channelInactive(ctx);
+        }
```
Comment (RongtongJin) [inline remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java:733]: In order to solve the issueeof timeout, is it worth scanning all items of responseTable?

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java (hunk 2 above)
Comment (qsrg) [reply remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java]: > Do not import xxx.*

Modified

File: remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java (hunk 3 above)
Comment (qsrg) [reply remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java]: I have optimized the structure, migrate wakeup logic to NettyRemotingAbstract. Similar to NettyRemotingAbstract#failFast, scan is required.

Comment (qsrg) [reply remoting/src/main/java/org/apache/rocketmq/remoting/netty/NettyRemotingClient.java]: > In order to solve the issueeof timeout, is it worth scanning all items of responseTable?

I have optimized the structure, migrate wakeup logic to NettyRemotingAbstract. Similar to NettyRemotingAbstract#failFast, scan is required.

**Pull request closed without merging**